modules/importer/src/runner/common/heartbeat.rs

```diff
@@ -0,0 +1,66 @@
+use super::Error;
+use crate::model::Importer;
+use sea_orm::{entity::*, prelude::*, QueryFilter};
+use time::OffsetDateTime;
+use tokio::{
+    task::JoinHandle,
+    time::{interval, Duration},
+};
+use trustify_common::db::Database;
+use trustify_entity::importer;
+
+pub struct Heart {
+    name: String,
+    handle: JoinHandle<()>,
+}
+
+impl Heart {
+    pub const RATE: Duration = Duration::from_secs(10);
+
+    pub fn new(importer: Importer, db: Database) -> Self {
+        let name = importer.name.clone();
+        let handle = tokio::spawn(async move {
+            let mut interval = interval(Heart::RATE);
+            let mut importer = importer;
+            loop {
+                interval.tick().await;
+                match Self::beat(&importer, &db).await {
+                    Ok(i) => {
+                        log::debug!("{}: {:#?}", i.name, i.data.progress);
+                        importer = i;
+                    }
+                    Err(e) => log::error!("Failed to send heartbeat for '{}': {e}", importer.name),
+                }
+            }
+        });
+        Self { name, handle }
+    }
+
+    // Updates the importer record with the current time, but only if
+    // the db matches our previous update, i.e. optimistic
+    // locking. Returns the updated record on success.
+    pub async fn beat(importer: &Importer, db: &Database) -> Result<Importer, Error> {
+        let t = OffsetDateTime::now_utc().unix_timestamp_nanos();
+        let model = importer::ActiveModel {
+            name: Set(importer.name.to_owned()),
+            heartbeat: Set(Some(t.into())),
+            ..Default::default()
+        };
+        use importer::Column::Heartbeat;
+        let lock = match importer.heartbeat {
+            Some(t) => Expr::col(Heartbeat).eq(Decimal::from_i128_with_scale(t, 0)),
+            None => Expr::col(Heartbeat).is_null(),
+        };
+        match importer::Entity::update(model).filter(lock).exec(db).await {
```

Comment:
I think the expectation is that this function returns a "mid air collision" in the case of a broken optimistic lock. But I can't seem to find the logic for that here.

Reply:
This function, `Heart::beat`, is called from two places. Initially, from the `server::import` function with the sole purpose of obtaining the lock. If it can't, it simply stops processing, and no heartbeat task is created.

Only after it acquires the lock is `Heart::new` called, which then creates the loop to continually call `Heart::beat`.

The "mid air collision" is never reported to the user. It's only used as a coordination mechanism.

Reply:
What happens if two instances call that function at roughly the same time? What happens, if one of them does not update that row? What happens if both update the row?

Reply:
There is no "roughly". They would have to call it at the exact same time, in nanoseconds. That's the only way both could update the row. And if that happens, they would each start running the same job. But I don't think it's going to. :smile:

Reply:
But if you think it might, we can easily add another column containing a unique id for the runner.

Reply:
What would be the outcome if no row would be updated by that update call?

Reply:
If the job has already started, and the heartbeat has been running for a while, and then encounters some error that prevents the update, that would then open the window for other instances to take on that job. Those instances would assume the runner previously executing the job had crashed without that heartbeat update.

Reply:
Then fields being added are starting to look a lot like the ones needed to track replica membership.  See:

https://github.com/trustification/trustify/compare/main...chirino:trustify:replica-manager

Reply:
I think the logic is simpler in this PR than what's in the replica manager, but they both share the same challenges of distributed computing, so not surprising they need similar fields.